go/engine/device_keyfinder.go

```diff
@@ -85,89 +78,81 @@ func (e *DeviceKeyfinder) UsersPlusDeviceKeys() map[keybase1.UID]UserPlusDeviceK
+	}
+	if err := e.addUser(eng.Result()); err != nil {
+		return err
+	}
 	return nil
 }
 
-func (e *DeviceKeyfinder) loadKeys(ctx *Context) error {
-	// get the device keys for all the users
-	for uid, uwk := range e.userMap {
-		var keys []keybase1.PublicKey
-		ckf := uwk.User.GetComputedKeyFamily()
-		if ckf != nil {
-			keys = ckf.ExportDeviceKeys()
-		}
+func (e *DeviceKeyfinder) hasUser(upk *keybase1.UserPlusKeys) bool {
+	_, ok := e.userMap[upk.Uid]
+	return ok
+}
 
-		if len(keys) == 0 {
-			return fmt.Errorf("User %s doesn't have a device key", uwk.User.GetName())
+func (e *DeviceKeyfinder) filterKeys(upk *keybase1.UserPlusKeys) error {
+	var keys []keybase1.PublicKey
+	for _, key := range upk.Keys {
+		if len(key.PGPFingerprint) != 0 {
+			continue
 		}
-		uwk.Keys = keys
-		e.userMap[uid] = uwk
+		if e.arg.NeedVerifyKeys && !libkb.KIDIsDeviceVerify(key.KID) {
+			continue
+		}
+		if e.arg.NeedEncryptKeys && !libkb.KIDIsDeviceEncrypt(key.KID) {
+			continue
+		}
+		keys = append(keys, key)
 	}
-
+	if len(keys) == 0 {
+		msg := fmt.Sprintf("User %s doesn't have any suitable keys", upk.Username)
+		return libkb.NoKeyError{Msg: msg}
+	}
+	upk.Keys = keys
 	return nil
 }
 
-func (e *DeviceKeyfinder) trackUser(ctx *Context, username string) error {
-	e.G().Log.Debug("tracking user %q", username)
-	arg := &TrackEngineArg{
-		Me:            e.arg.Me,
-		UserAssertion: username,
-		Options:       e.arg.TrackOptions,
+func (e *DeviceKeyfinder) addUser(ir *keybase1.Identify2Res) error {
+
+	if ir == nil {
+		return fmt.Errorf("Null result from Identify2")
 	}
-	eng := NewTrackEngine(arg, e.G())
-	err := RunEngine(eng, ctx)
-	// ignore self track errors
-	if _, ok := err.(libkb.SelfTrackError); ok {
-		e.addUser(e.arg.Me, false)
+	upk := &ir.Upk
+
+	if e.hasUser(upk) {
 		return nil
 	}
-	if err != nil {
-		return err
-	}
-	e.addUser(eng.User(), true)
-	return nil
-}
 
-// PC: maybe we need to bring the TrackUI back for the
-// context...so that this one can use an IdentifyUI and trackUser
-// can use a TrackUI...
-func (e *DeviceKeyfinder) identifyUser(ctx *Context, user string) error {
-	arg := NewIdentifyArg(user, false, false)
-	eng := NewIdentify(arg, e.G())
-	if err := RunEngine(eng, ctx); err != nil {
-		return err
+	if e.arg.Self != nil && e.arg.Self.GetUID().Equal(upk.Uid) {
+		e.G().Log.Debug("skipping self in DevicekeyFinder (uid=%s)", upk.Uid)
+		return nil
 	}
-	e.addUser(eng.User(), false)
-	return nil
-}
 
-func (e *DeviceKeyfinder) hasUser(user *libkb.User) bool {
-	_, ok := e.userMap[user.GetUID()]
-	return ok
-}
-
-func (e *DeviceKeyfinder) addUser(user *libkb.User, tracked bool) {
-	if e.hasUser(user) {
-		return
+	if err := e.filterKeys(upk); err != nil {
+		return err
 	}
 
 	index := len(e.userMap)
-	e.userMap[user.GetUID()] = UserPlusDeviceKeys{
-		User:      user,
-		Index:     index,
-		IsTracked: tracked,
+	e.userMap[upk.Uid] = UserPlusDeviceKeys{
+		Upk:   upk,
+		Index: index,
```

Comment:
I might have forgotten to remove it when copying this stuff over from the PGPKeyfinder! I think we can remove this if nothing breaks...